[awsv2] Panic when no parent segment is present

A panic error is raised when trying to retrieve a secret value, without having a parent segment, even with `AWS_XRAY_CONTEXT_MISSING` set to `"LOG_ERROR"` or `"IGNORE_ERROR"`.

This is a common use case, when having a Lambda function and needing to retrieve a secret value just once during its `init()` block. During **Initialization**, there's no parent segment, so setting `AWS_XRAY_CONTEXT_MISSING=LOG_ERROR` or `AWS_XRAY_CONTEXT_MISSING=IGNORE_ERROR` is needed to avoid the Lambda function from failing indefinitely.

## Reproducible code example

`go.mod`
```go
module example.com/test-aws-sdk-xray

go 1.16

require (
	github.com/aws/aws-sdk-go-v2 v1.7.0
	github.com/aws/aws-sdk-go-v2/config v1.4.1
	github.com/aws/aws-sdk-go-v2/service/secretsmanager v1.4.0
	github.com/aws/aws-xray-sdk-go v1.5.1-0.20210701195726-f36e5263a798
)
```

`main.go`
```go
package main

import (
	"context"
	"log"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/secretsmanager"
	"github.com/aws/aws-xray-sdk-go/instrumentation/awsv2"
)

func main() {
	ctx := context.TODO()

	cfg, err := config.LoadDefaultConfig(ctx)
	if err != nil {
		log.Fatal(err)
	}

	// Instrument AWS SDK v2.
	awsv2.AWSV2Instrumentor(&cfg.APIOptions)

	client := secretsmanager.NewFromConfig(cfg)

	resp, err := client.GetSecretValue(
		ctx,
		&secretsmanager.GetSecretValueInput{
			SecretId: aws.String("<SECRET_ARN>"),
		},
	)
	if err != nil {
		log.Fatal(err)
	}

	log.Println("Secret content:", *resp.SecretString)
}
```

## Error message

Specifying `AWS_XRAY_CONTEXT_MISSING` as `LOG_ERROR` (**unexpected panic**)
```
$ AWS_XRAY_CONTEXT_MISSING=LOG_ERROR go run ./main.go
2021-07-06T18:09:18-03:00 [ERROR] Suppressing AWS X-Ray context missing panic: failed to begin subsegment named 'Secrets Manager': segment cannot be found.
panic: interface conversion: interface {} is nil, not *secretsmanager.GetSecretValueOutput

goroutine 1 [running]:
github.com/aws/aws-sdk-go-v2/service/secretsmanager.(*Client).GetSecretValue(0xc0003d8210, 0xd76dd0, 0xc00019c008, 0xc0001a89d8, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc0001d5ac0)
        /home/mike/.go/pkg/mod/github.com/aws/aws-sdk-go-v2/service/secretsmanager@v1.4.0/api_op_GetSecretValue.go:43 +0x1f5
main.main()
        /home/mike/dev/test-aws-sdk-xray/main.go:26 +0x29b
exit status 2
```

Specifying `AWS_XRAY_CONTEXT_MISSING` as `IGNORE_ERROR` (**unexpected panic**)
```
$ AWS_XRAY_CONTEXT_MISSING=IGNORE_ERROR go run ./main.go
panic: interface conversion: interface {} is nil, not *secretsmanager.GetSecretValueOutput

goroutine 1 [running]:
github.com/aws/aws-sdk-go-v2/service/secretsmanager.(*Client).GetSecretValue(0xc000356210, 0xd76dd0, 0xc00003e118, 0xc00000e9f0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc00009bb00)
        /home/mike/.go/pkg/mod/github.com/aws/aws-sdk-go-v2/service/secretsmanager@v1.4.0/api_op_GetSecretValue.go:43 +0x1f5
main.main()
        /home/mike/dev/test-aws-sdk-xray/main.go:26 +0x29b
exit status 2
```

## Additional context

* Running the code with `AWS_XRAY_SDK_DISABLED=TRUE` avoids the panic and returns correctly.
* Not specifying `AWS_XRAY_CONTEXT_MISSING` (or setting it to `"RUNTIME_ERROR"`), generates an expected panic.

* The following code with the AWS SDK v1, also works as expected, without any issues:

`go.mod`
```go
module example.com/test-aws-sdk-xray-v1

go 1.16

require (
	github.com/aws/aws-sdk-go v1.17.12
	github.com/aws/aws-xray-sdk-go v1.5.1-0.20210701195726-f36e5263a798
)
```

`main.go`
```go
package main

import (
	"context"
	"log"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/secretsmanager"
	"github.com/aws/aws-xray-sdk-go/xray"
)

func main() {
	ctx := context.TODO()

	sess := session.Must(session.NewSession())
	secrets := secretsmanager.New(sess)

	// Instrument AWS SDK v1.
	xray.AWS(secrets.Client)

	resp, err := secrets.GetSecretValueWithContext(
		ctx,
		&secretsmanager.GetSecretValueInput{
			SecretId: aws.String("<SECRET_ARN>"),
		},
	)
	if err != nil {
		log.Fatal(err)
	}

	log.Println("Secret content:", *resp.SecretString)
}
```

--- a/instrumentation/awsv2/awsv2.go
+++ b/instrumentation/awsv2/awsv2.go
@@ -28,7 +28,7 @@ func initializeMiddlewareAfter(stack *middleware.Stack) error {
 		// Start the subsegment
 		ctx, subseg := xray.BeginSubsegment(ctx, serviceName)
 		if subseg == nil {
-			return
+			return next.HandleInitialize(ctx, in)
 		}
 		subseg.Namespace = "aws"
 		subseg.GetAWS()["region"] = v2Middleware.GetRegion(ctx)
@@ -52,7 +52,11 @@ func deserializeMiddleware(stack *middleware.Stack) error {
 		ctx context.Context, in middleware.DeserializeInput, next middleware.DeserializeHandler) (
 		out middleware.DeserializeOutput, metadata middleware.Metadata, err error) {
 
-		subseg := ctx.Value(awsV2SubsegmentKey{}).(*xray.Segment)
+		subseg, ok := ctx.Value(awsV2SubsegmentKey{}).(*xray.Segment)
+		if !ok {
+			return next.HandleDeserialize(ctx, in)
+		}
+
 		in.Request.(*smithyhttp.Request).Header.Set(xray.TraceIDHeaderKey, subseg.DownstreamHeader().String())
 
 		out, metadata, err = next.HandleDeserialize(ctx, in)

--- a/instrumentation/awsv2/awsv2_test.go
+++ b/instrumentation/awsv2/awsv2_test.go
@@ -21,6 +21,7 @@ import (
 	"github.com/aws/aws-sdk-go-v2/aws"
 	"github.com/aws/aws-sdk-go-v2/service/route53"
 	"github.com/aws/aws-sdk-go-v2/service/route53/types"
+	"github.com/aws/aws-xray-sdk-go/strategy/ctxmissing"
 	"github.com/aws/aws-xray-sdk-go/xray"
 )
 
@@ -147,6 +148,95 @@ func TestAWSV2(t *testing.T) {
 					t.Errorf("expected request id to be %s, got %s", e, a)
 				}
 			}
+		})
+		time.Sleep(1 * time.Second)
+	}
+}
+
+func TestAWSV2WithoutSegment(t *testing.T) {
+	cases := map[string]struct {
+		responseStatus int
+		responseBody   []byte
+	}{
+		"fault response": {
+			responseStatus: 500,
+			responseBody: []byte(`<?xml version="1.0" encoding="UTF-8"?>
+		<InvalidChangeBatch xmlns="https://route53.amazonaws.com/doc/2013-04-01/">
+		<Messages>
+		  <Message>Tried to create resource record set duplicate.example.com. type A, but it already exists</Message>
+		</Messages>
+		<RequestId>b25f48e8-84fd-11e6-80d9-574e0c4664cb</RequestId>
+		</InvalidChangeBatch>`),
+		},
+
+		"error response": {
+			responseStatus: 404,
+			responseBody: []byte(`<?xml version="1.0"?>
+		<ErrorResponse xmlns="http://route53.amazonaws.com/doc/2016-09-07/">
+		<Error>
+		  <Type>Sender</Type>
+		  <Code>MalformedXML</Code>
+		  <Message>1 validation error detected: Value null at 'route53#ChangeSet' failed to satisfy constraint: Member must not be null</Message>
+		</Error>
+		<RequestId>1234567890A</RequestId>
+		</ErrorResponse>
+		`),
+		},
+
+		"success response": {
+			responseStatus: 200,
+			responseBody: []byte(`<?xml version="1.0" encoding="UTF-8"?>
+		<ChangeResourceRecordSetsResponse>
+			<ChangeInfo>
+			<Comment>mockComment</Comment>
+			<Id>mockID</Id>
+		</ChangeInfo>
+		</ChangeResourceRecordSetsResponse>`),
+		},
+	}
+
+	for name, c := range cases {
+		server := httptest.NewServer(http.HandlerFunc(
+			func(w http.ResponseWriter, r *http.Request) {
+				w.WriteHeader(c.responseStatus)
+				_, err := w.Write(c.responseBody)
+				if err != nil {
+					t.Fatal(err)
+				}
+			}))
+		defer server.Close()
+
+		t.Run(name, func(t *testing.T) {
+			// Ignore errors when segment cannot be found.
+			ctx, err := xray.ContextWithConfig(
+				context.Background(),
+				xray.Config{ContextMissingStrategy: ctxmissing.NewDefaultIgnoreErrorStrategy()},
+			)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			svc := route53.NewFromConfig(aws.Config{
+				EndpointResolver: aws.EndpointResolverFunc(func(service, region string) (aws.Endpoint, error) {
+					return aws.Endpoint{
+						URL:         server.URL,
+						SigningName: "route53",
+					}, nil
+				}),
+				Retryer: func() aws.Retryer {
+					return aws.NopRetryer{}
+				},
+			})
+
+			_, _ = svc.ChangeResourceRecordSets(ctx, &route53.ChangeResourceRecordSetsInput{
+				ChangeBatch: &types.ChangeBatch{
+					Changes: []types.Change{},
+					Comment: aws.String("mock"),
+				},
+				HostedZoneId: aws.String("zone"),
+			}, func(options *route53.Options) {
+				AWSV2Instrumentor(&options.APIOptions)
+			})
 		})
 		time.Sleep(1 * time.Second)
 	}

